blog/2025-02-03/index.md

```diff
@@ -0,0 +1,67 @@
+
+Fast-forward to 2025. These hacks no longer work with the latest release of Buildbot, rendering our soft fork stuck on a version from 2021. For whatever reason, updating GitHub CI status stopped to work well, allowing people to merge untested code changes. It also requires periodical restarts because it grinds to a halt for no particular reason, dropping new builds in the meantime. The worst offense: I never liked it.
+
+Interlude: Equinix Metal née Packet has been sponsoring our heavy-lifting servers doing actual building for the past 5 years. Unfortunately, they are shutting down, meaning we need to move out by the end of April 2025.
+
+This perfect storm means we need to effectively re-architecture build infrastructure from scratch.
+
+## webhook-proxy
+
+Let's start from improving Buildbot reliability where it falls short. While GitHub shows delivery status of webhook events, and even provides a button to trigger another attempt, there's no public API exposing this data.
+
+As there's no way I'm fixing Buildbot itself, I decided we need a middleman which will take care of redelivering if Buildbot's webhook endpoint responded with non-200 response.
+
+[webhook-proxy][webhook-proxy] is a simple Python service which accepts payloads from GitHub, marks new pull requests as pending CI, then forwards unchanged events to Buildbot. Should Buildbot have a hiccup, it will open a circuit breaker and retry with exponential back off until it eventually succeeds.
+
+It ain't much, but it's honest work.
+
+## justpak
+
+An important piece of the future Buildbot replacement is being agnostic of its CI/CD implementation. The idea is to provide only a business logic service, while actual building is delegated elsewhere. As such, it requires some relatively generic tool to execute tasks instead of maintaining two or more pipeline definitions for whatever is the CI/CD solution of the decade.
+
+After evaluating modern `make` replacements, I settled on [just][just]. It seems simple enough for executing external commands, and has a nifty way of defining recipes with a specific shebang from the get-go, so it's all single file even when using a mix of Python and Bash.
+
+I copied all build steps done by the Buildbot workers to specific recipes to have a single entry point. Then I started doing back flips with GitHub Actions to see if this is a viable way in the first place; turns out it is, although not without some hurdles.
+
+While GitHub Actions have a native way of executing all steps inside a Docker container, the host running said container is full of bloat and barely any free disk space. [bbhtt][bbhtt] suggested how to remove unneeded files, which also meant each build step is prefixed with `docker run` as we want to re-use existing [flatpak-builder-lint][flatpak-builder-lint] image instead of meddling with Ubuntu.
```

Comment:
```suggestion
While GitHub Actions have a native way of executing all steps inside a Docker container, the host running said container is full of bloat and barely has any free disk space. [bbhtt][bbhtt] suggested how to remove unneeded files, which also meant each build step is prefixed with `docker run` as we want to re-use existing [flatpak-builder-lint][flatpak-builder-lint] image instead of meddling with Ubuntu.
```